Code/GraphMol/FileParsers/PNGParser.cpp

```diff
@@ -0,0 +1,190 @@
+                                << std::endl;
+      }
+      res[key] = value;
+    }
+    inStream.seekg(beginBlock);
+    inStream.ignore(blockLen + 4);  // the extra 4 bytes are the CRC
+  }
+
+  return res;
+};
+
+std::string addMetadataToPNGStream(
+    std::istream &inStream,
+    const std::map<std::string, std::string> &metadata) {
+  // confirm that it's a PNG file:
+  if (!checkPNGHeader(inStream)) {
+    throw FileParseException("PNG header not recognized");
+  }
+  std::stringstream res;
+  // write the header
+  for (auto byte : pngHeader) {
+    res << byte;
+  }
+
+  // copy over everything up to IEND
+  bool foundEnd = false;
+  std::uint32_t finalCRC;
+  while (inStream) {
+    std::uint32_t blockLen;
+    inStream.read((char *)&blockLen, sizeof(blockLen));
+    char bytes[4];
+    inStream.read(bytes, 4);
+    if (bytes[0] == 'I' && bytes[1] == 'E' && bytes[2] == 'N' &&
+        bytes[3] == 'D') {
+      foundEnd = true;
+      inStream.read((char *)&finalCRC, sizeof(finalCRC));
+      break;
+    }
+    res.write((char *)&blockLen, sizeof(blockLen));
+    res.write(bytes, 4);
+    // PNG is big endian, make sure we handle the order correctly
+    blockLen = EndianSwapBytes<BIG_ENDIAN_ORDER, HOST_ENDIAN_ORDER>(blockLen);
+    std::string block(blockLen + 4, 0);
+    inStream.read((char *)&block.front(),
+                  blockLen + 4);  // the extra 4 bytes are the CRC
+    res.write(block.c_str(), blockLen + 4);
+  }
+  if (!foundEnd) {
+    throw FileParseException("did not find IEND block in PNG");
+  }
+
+  // write out the metadata:
+  for (const auto pr : metadata) {
+    std::stringstream blk;
+    blk.write("tEXt", 4);
+    // write the name along with a zero
+    blk.write(pr.first.c_str(), pr.first.size() + 1);
+    blk.write(pr.second.c_str(), pr.second.size());
+    auto blob = blk.str();
+    std::uint32_t blksize =
+        blob.size() - 4;  // we don't include the tag in the size;
+    boost::crc_32_type crc;
+    crc.process_bytes((void const *)blob.c_str(), blob.size());
+    std::uint32_t crcVal = crc.checksum();
+    // PNG is big endian, make sure we handle the order correctly
+    blksize = EndianSwapBytes<HOST_ENDIAN_ORDER, BIG_ENDIAN_ORDER>(blksize);
+
+    res.write((char *)&blksize, sizeof(blksize));
+    res.write(blob.c_str(), blob.size());
+    // PNG is big endian, make sure we handle the order correctly
+    crcVal = EndianSwapBytes<HOST_ENDIAN_ORDER, BIG_ENDIAN_ORDER>(crcVal);
+    res.write((char *)&crcVal, sizeof(crcVal));
+  }
+
+  // write out the IEND block
+  std::uint32_t blksize = 0;
+  res.write((char *)&blksize, sizeof(blksize));
+
+  const char *endTag = "IEND";
+  res.write(endTag, 4);
+  res.write((char *)&finalCRC, sizeof(finalCRC));
+  return res.str();
+}
+
+std::string addMolToPNGStream(std::istream &iStream, const ROMol &mol) {
+  std::string smi = MolToCXSmiles(mol);
+  std::map<std::string, std::string> metadata;
```

Comment:
Does this add the molecule name/title as well?

Reply:
@greglandrum I think adding the _Name property to the metadata is quite valuable.  I would use that pretty heavily I think.  Perhaps it could be an optional std::vector<string> of data properties to add to the metadata.  I could see adding  pIC50s and other bio data.